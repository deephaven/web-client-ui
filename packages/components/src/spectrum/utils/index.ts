export * from './itemUtils';
export * from './itemWrapperUtils';
export * from './propsUtils';
export * from './themeUtils';
export * from './useRenderNormalizedItem';
export * from './useStringifiedMultiSelection';
export * from './useStringifiedSelection';
