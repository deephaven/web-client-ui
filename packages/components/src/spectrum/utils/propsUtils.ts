import type { AriaLabelingProps, StyleProps } from '@react-types/shared';

/**
 * Separate props for Spectrum components into AriaLabelingProps, StyleProps, and
 * any remaining props.
 * @param props The props to separate
 * @returns The separated props
 */
export function separateSpectrumProps<T extends AriaLabelingProps & StyleProps>(
  props: T
): {
  ariaLabelProps: AriaLabelingProps;
  styleProps: StyleProps;
  componentProps: Omit<T, keyof (AriaLabelingProps & StyleProps)>;
} {
  const {
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledby,
    'aria-describedby': ariaDescribedby,
    'aria-details': ariaHidden,

    marginX,
    marginY,
    width,
    height,
    minWidth,
    minHeight,
    maxWidth,
    maxHeight,
    flex,
    flexGrow,
    flexShrink,
    flexBasis,
    justifySelf,
    alignSelf,
    order,
    gridArea,
    gridColumn,
    gridRow,
    gridColumnStart,
    gridColumnEnd,
    gridRowStart,
    gridRowEnd,
    position,
    zIndex,
    top,
    bottom,
    start,
    end,
    left,
    right,
    isHidden,
    ...componentProps
  } = props;

  return {
    ariaLabelProps: {
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledby,
      'aria-describedby': ariaDescribedby,
      'aria-details': ariaHidden,
    },
    styleProps: {
      marginX,
      marginY,
      width,
      height,
      minWidth,
      minHeight,
      maxWidth,
      maxHeight,
      flex,
      flexGrow,
      flexShrink,
      flexBasis,
      justifySelf,
      alignSelf,
      order,
      gridArea,
      gridColumn,
      gridRow,
      gridColumnStart,
      gridColumnEnd,
      gridRowStart,
      gridRowEnd,
      position,
      zIndex,
      top,
      bottom,
      start,
      end,
      left,
      right,
      isHidden,
    },
    componentProps: componentProps as Omit<
      T,
      keyof (AriaLabelingProps & StyleProps)
    >,
  };
}

export default separateSpectrumProps;
