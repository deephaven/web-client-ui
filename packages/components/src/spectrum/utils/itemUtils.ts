import { Key, ReactElement, ReactNode } from 'react';
import type { ItemRenderer } from '@react-types/shared';
import { isElementOfType } from '@deephaven/react-hooks';
import { ensureArray, KeyedItem, SelectionT } from '@deephaven/utils';
import {
  Item,
  ItemElementOrPrimitive,
  ItemProps,
  Section,
  SectionProps,
} from '../shared';
import { PopperOptions } from '../../popper';
import { Text } from '../Text';
import ItemContent from '../ItemContent';

/**
 * `Item.textValue` prop needs to be a non-empty string for accessibility
 * purposes. This is not displayed in the UI.
 */
export const ITEM_EMPTY_STRING_TEXT_VALUE = 'Empty';

/**
 * React Spectrum <Section> supports an `ItemRenderer` function as a child. The
 * DH picker makes use of this internally, but we don't want to support it as
 * an incoming prop.
 */
type SectionPropsNoItemRenderer<T> = Omit<SectionProps<T>, 'children'> & {
  children: Exclude<SectionProps<T>['children'], ItemRenderer<T>>;
};

export type ItemElement<T = unknown> = ReactElement<ItemProps<T>>;
export type SectionElement<T = unknown> = ReactElement<
  SectionPropsNoItemRenderer<T>
>;

export type ItemOrSection<T = unknown> =
  | ItemElementOrPrimitive<T>
  | SectionElement<T>;

// Picker uses `icon` slot. ListView can use `image` or `illustration` slots.
// https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/picker/src/Picker.tsx#L194
// https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/list/src/ListViewItem.tsx#L266-L267
export type ItemIconSlot = 'icon' | 'image' | 'illustration';

/**
 * Augment the Spectrum selection key type to include boolean values.
 * Spectrum collection components already supports this, but the built in types
 * don't reflect it.
 */
export type ItemKey = Key | boolean;

export type ItemSelection = SelectionT<ItemKey>;

/**
 * Augment the Spectrum selection change handler type to include boolean keys.
 * Spectrum components already supports this, but the built in types don't
 * reflect it.
 */
export type ItemSelectionChangeHandler = (key: ItemKey) => void;

export interface MultipleItemSelectionProps {
  selectedKeys?: 'all' | Iterable<ItemKey>;
  defaultSelectedKeys?: 'all' | Iterable<ItemKey>;
  disabledKeys?: Iterable<ItemKey>;

  /**
   * Handler that is called when the selection change.
   * Note that under the hood, this is just an alias for Spectrum's
   * `onSelectionChange`. We are renaming for better consistency with other
   * components.
   */
  onChange?: (keys: ItemSelection) => void;

  /**
   * Handler that is called when the selection changes.
   * @deprecated Use `onChange` instead
   */
  onSelectionChange?: (keys: ItemSelection) => void;
}

export interface NormalizedItemData {
  key?: ItemKey;
  content: ReactNode;
  description?: ReactNode;
  icon?: ReactNode;
  textValue: string | undefined;
}

export interface NormalizedSectionData {
  key?: Key;
  title?: ReactNode;
  items: NormalizedItem[];
}

/**
 * Spectrum collection components support a variety of item types, including
 * strings, numbers, booleans, and more complex React elements. This type
 * represents a normalized form to make rendering items simpler and keep the
 * logic of transformation in separate util methods. It also adheres to the
 * `KeyedItem` interface to be compatible with Windowed data utils
 * (e.g. `useViewportData`).
 */
export type NormalizedItem = KeyedItem<NormalizedItemData, ItemKey>;

export type NormalizedSection = KeyedItem<NormalizedSectionData, Key>;

export type NormalizedItemOrSection<TItemOrSection extends ItemOrSection> =
  TItemOrSection extends SectionElement ? NormalizedSection : NormalizedItem;

export type TooltipOptions = { placement: PopperOptions['placement'] };

/**
 * DH wrappers of Spectrum collection components use a normalized item that
 * includes a `key` prop and an optional `item` prop. This is mostly to support
 * Windowed data where items are created before their data has been loaded (data
 * gets set in the `item` prop). If data has loaded, return its `key`. If not,
 * return the top-level `key` on the normalized item.
 * @param item The normalized item or section
 * @returns The `key` of the item or section
 */
export function getItemKey<
  TItem extends NormalizedItem | NormalizedSection,
  TKey extends TItem extends NormalizedItem
    ? ItemKey
    : TItem extends NormalizedSection
    ? Key
    : undefined,
>(item: TItem | null | undefined): TKey {
  return (item?.item?.key ?? item?.key) as TKey;
}

/**
 * Determine Item `textValue` based on the `textValue` prop or primitive children
 * value.
 * @param item The item to get the text value for
 * @returns The text value of the item
 */
export function getItemTextValue<T>(item: ItemElement<T>): string | undefined {
  if (item.props.textValue == null) {
    const itemKeyStr = item.key == null ? undefined : String(item.key);
    return ['string', 'boolean', 'number'].includes(typeof item.props.children)
      ? String(item.props.children)
      : itemKeyStr;
  }

  return item.props.textValue === ''
    ? ITEM_EMPTY_STRING_TEXT_VALUE
    : item.props.textValue;
}

/**
 * Get the position of the item with the given selected key in a list of items.
 * @param items The items to search
 * @param itemHeight The height of each item
 * @param selectedKey The key of the selected item
 * @param topOffset The top offset of the list
 * @returns The position of the selected item or the top offset if not found
 */
export async function getPositionOfSelectedItemElement<
  TKey extends string | number | boolean | undefined,
>({
  items,
  itemHeight,
  selectedKey,
  topOffset,
}: {
  items: ItemElement[];
  selectedKey: TKey | null | undefined;
  itemHeight: number;
  topOffset: number;
}): Promise<number> {
  let position = topOffset;

  if (selectedKey == null) {
    return position;
  }

  for (let i = 0; i < items.length; i += 1) {
    const item = items[i];
    if (item.key === selectedKey) {
      return position;
    }

    position += itemHeight;
  }

  return topOffset;
}

/**
 * Determine if a node is a Section element.
 * @param node The node to check
 * @returns True if the node is a Section element
 */
export function isSectionElement<T>(
  node: ReactNode
): node is SectionElement<T> {
  return isElementOfType(node, Section);
}

/**
 * Determine if a node is an Item element.
 * @param node The node to check
 * @returns True if the node is an Item element
 */
export function isItemElement<T>(node: ReactNode): node is ItemElement<T> {
  return isElementOfType(node, Item);
}

/**
 * Determine if a node is an Item element containing a child `Text` element with
 * a `slot` prop set to `description`.
 * @param node The node to check
 * @returns True if the node is an Item element with a description
 */
export function isItemElementWithDescription<T>(
  node: ReactNode
): node is ReactElement<ItemProps<T>> {
  if (!isItemElement(node)) {
    return false;
  }

  // If children are wrapped in `ItemContent`, go down 1 level
  const children = isElementOfType(node.props.children, ItemContent)
    ? node.props.children.props.children
    : node.props.children;

  const childrenArray = ensureArray(children);

  const result = childrenArray.some(
    child => isElementOfType(child, Text) && child.props?.slot === 'description'
  );

  return result;
}

/**
 * Determine if a node is an array containing normalized items or sections with
 * keys. Note that this only checks the first node in the array.
 * @param node The node to check
 * @returns True if the node is a normalized item or section with keys array
 */
export function isNormalizedItemsWithKeysList<
  TItemOrSection extends ItemOrSection,
>(
  node:
    | TItemOrSection
    | TItemOrSection[]
    | NormalizedItemOrSection<TItemOrSection>[]
): node is NormalizedItemOrSection<TItemOrSection>[] {
  if (!Array.isArray(node)) {
    return false;
  }

  if (node.length === 0) {
    return true;
  }

  return (
    !isItemOrSection(node[0]) && typeof node[0] === 'object' && 'key' in node[0]
  );
}

/**
 * Determine if an object is a normalized section.
 * @param maybeNormalizedSection The object to check
 * @returns True if the object is a normalized section
 */
export function isNormalizedSection(
  maybeNormalizedSection: NormalizedItem | NormalizedSection
): maybeNormalizedSection is NormalizedSection {
  return (
    maybeNormalizedSection.item != null &&
    'items' in maybeNormalizedSection.item
  );
}

/**
 * Determine if a node is an item or section. Valid types include strings,
 * numbers, booleans, Item elements, and Section elements.
 * @param node The node to check
 * @returns True if the node is an item or section
 */
export function isItemOrSection(node: ReactNode): node is ItemOrSection {
  return (
    typeof node === 'string' ||
    typeof node === 'number' ||
    typeof node === 'boolean' ||
    isItemElement(node) ||
    isSectionElement(node)
  );
}

/**
 * Returns a TooltipOptions object or null if options is false or null.
 * @param options Tooltip options
 * @param placement Default placement for the tooltip if `options` is set
 * explicitly to `true`
 * @returns TooltipOptions or null
 */
export function normalizeTooltipOptions(
  options?: boolean | TooltipOptions | null,
  placement: TooltipOptions['placement'] = 'right'
): TooltipOptions | null {
  if (options == null || options === false) {
    return null;
  }

  if (options === true) {
    return { placement };
  }

  return options;
}

/**
 * Convert a selection of `ItemKey`s to a selection of strings.
 * @param itemKeys The selection of `ItemKey`s
 * @returns The selection of strings
 */
export function itemSelectionToStringSet<
  TKeys extends 'all' | Iterable<ItemKey> | undefined,
  TResult extends TKeys extends 'all'
    ? 'all'
    : TKeys extends Iterable<ItemKey>
    ? Set<string>
    : undefined,
>(itemKeys: TKeys): TResult {
  if (itemKeys == null || itemKeys === 'all') {
    return itemKeys as undefined | 'all' as TResult;
  }

  return new Set([...itemKeys].map(String)) as TResult;
}
