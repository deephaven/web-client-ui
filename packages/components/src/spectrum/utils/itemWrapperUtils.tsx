import { cloneElement, ReactNode } from 'react';
import { Item } from '@adobe/react-spectrum';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { dh as dhIcons } from '@deephaven/icons';
import { isElementOfType } from '@deephaven/react-hooks';
import { ensureArray, NON_BREAKING_SPACE } from '@deephaven/utils';
import {
  isItemElement,
  isSectionElement,
  ItemElement,
  ItemIconSlot,
  ItemOrSection,
  ITEM_EMPTY_STRING_TEXT_VALUE,
  SectionElement,
  TooltipOptions,
} from './itemUtils';
import { ItemContent } from '../ItemContent';
import { Icon } from '../icons';
import { Text } from '../Text';

/**
 * If the given content is a string, wrap it in an Icon component. Otherwise,
 * return the original content. If the key is not found in the dhIcons object,
 * the vsBlank icon will be used.
 * @param maybeIconKey The content to wrap
 * @param slot The slot to use for the Icon component
 * @returns The wrapped content or original content if not a string
 */
export function wrapIcon(
  maybeIconKey: ReactNode,
  slot: ItemIconSlot
): ReactNode {
  // eslint-disable-next-line no-param-reassign
  maybeIconKey = maybeIconKey ?? '';

  if (typeof maybeIconKey !== 'string') {
    return maybeIconKey;
  }

  return (
    <Icon slot={slot}>
      <FontAwesomeIcon icon={dhIcons[maybeIconKey] ?? dhIcons.vsBlank} />
    </Icon>
  );
}

/**
 * Ensure all primitive children are wrapped in `Item` elements and that all
 * `Item` element content is wrapped in `ItemContent` elements to handle text
 * overflow consistently and to support tooltips.
 * @param itemsOrSections The items or sections to wrap
 * @param tooltipOptions The tooltip options to use when wrapping items
 * @returns The wrapped items or sections
 */
export function wrapItemChildren<T>(
  itemsOrSections: ItemOrSection<T> | ItemOrSection<T>[],
  tooltipOptions: TooltipOptions | null
): ItemElement<T> | SectionElement<T> | (ItemElement<T> | SectionElement<T>)[] {
  const itemsOrSectionsArray = ensureArray(itemsOrSections);

  const result = itemsOrSectionsArray.map(item => {
    if (isItemElement(item)) {
      // Item content is already wrapped
      if (isElementOfType(item.props.children, ItemContent)) {
        return item;
      }

      const key = item.key ?? item.props.textValue;
      const textValue =
        item.props.textValue === ''
          ? ITEM_EMPTY_STRING_TEXT_VALUE
          : item.props.textValue;

      // Wrap in `ItemContent` so we can support tooltips and handle text
      // overflow
      return cloneElement(item, {
        ...item.props,
        key,
        textValue,
        children: (
          <ItemContent tooltipOptions={tooltipOptions}>
            {item.props.children}
          </ItemContent>
        ),
      });
    }

    if (isSectionElement(item)) {
      return cloneElement(item, {
        ...item.props,
        key:
          item.key ??
          (typeof item.props.title === 'string' ? item.props.title : undefined),
        children: wrapItemChildren<T>(item.props.children, tooltipOptions),
      });
    }

    const text = String(item);
    const textValue = text === '' ? ITEM_EMPTY_STRING_TEXT_VALUE : text;

    return (
      <Item key={text} textValue={textValue}>
        <ItemContent tooltipOptions={tooltipOptions}>{text}</ItemContent>
      </Item>
    );
  });

  // Keep consistent with original data structure as array or single node
  return Array.isArray(itemsOrSections) ? result : result[0];
}

/**
 * If the given content is a primitive type, wrap it in a Text component.
 * @param content The content to wrap
 * @param slot The slot to use for the Text component
 * @returns The wrapped content or original content if not a primitive type
 */
export function wrapPrimitiveWithText(
  content?: ReactNode,
  slot?: string
): ReactNode {
  // eslint-disable-next-line no-param-reassign
  content = content ?? '';

  if (['string', 'boolean', 'number'].includes(typeof content)) {
    return (
      <Text slot={slot}>
        {/* Non-breaking space is needed to avoid the Text element's height 
        collapsing when content is empty */}
        {content === '' ? NON_BREAKING_SPACE : String(content)}
      </Text>
    );
  }

  return content;
}
