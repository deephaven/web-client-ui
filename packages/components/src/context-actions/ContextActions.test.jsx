import React from 'react';
import TestRenderer from 'react-test-renderer';
import { TestUtils } from '@deephaven/utils';
import { ContextMenuRoot, ContextActionUtils } from '.';

jest.mock('react-transition-group', () => ({
  CSSTransition: 'cssTransition',
  TransitionGroup: 'transitionGroup',
}));

const TEST_MENU_1 = [
  { title: 'Test1' },
  { title: 'Test2' },
  { title: 'Test3' },
];

function contextMenuMock(mock, element) {
  if (element.props && element.props.className) {
    if (element.props.className.indexOf('context-menu') !== -1) {
      return mock;
    }
  }

  return null;
}

function DEFAULT_MOCK(mockParent = true) {
  return {
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    getBoundingClientRect: jest.fn(() => ({
      top: 0,
      left: 0,
      width: 0,
      height: 0,
    })),
    focus: jest.fn(),
    offsetParent: mockParent ? DEFAULT_MOCK(false) : null,
    parentElement: mockParent ? DEFAULT_MOCK(false) : null,
    setAttribute: jest.fn(),
  };
}

it('renders without crashing', () => {
  const mock = DEFAULT_MOCK();

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  expect(mock.parentElement.addEventListener.mock.calls.length).toBe(1);
  expect(mock.parentElement.addEventListener.mock.calls[0][0]).toBe(
    'contextmenu'
  );
  expect(mock.parentElement.addEventListener.mock.calls[0][1]).not.toBeNull();
  expect(mock.parentElement.removeEventListener.mock.calls.length).toBe(0);

  expect(tree).toMatchSnapshot();

  tree.unmount();

  expect(mock.parentElement.addEventListener.mock.calls.length).toBe(1);
  expect(mock.parentElement.removeEventListener.mock.calls.length).toBe(1);
});

it('handles the context menu event on the parent, renders an array of menu items', () => {
  const mock = DEFAULT_MOCK();

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  expect(mock.parentElement.addEventListener.mock.calls.length).toBe(1);
  expect(mock.parentElement.addEventListener.mock.calls[0][0]).toBe(
    'contextmenu'
  );
  expect(mock.parentElement.addEventListener.mock.calls[0][1]).not.toBeNull();
  expect(mock.parentElement.removeEventListener.mock.calls.length).toBe(0);

  expect(tree.root.findAllByType('button').length).toBe(0);

  expect(tree).toMatchSnapshot();

  const mockEvent = { preventDefault: jest.fn(), contextActions: TEST_MENU_1 };
  const handleContextMenu =
    mock.parentElement.addEventListener.mock.calls[0][1];
  handleContextMenu(mockEvent);

  expect(tree.root.findAllByType('button').length).toBe(TEST_MENU_1.length);

  tree.unmount();

  expect(mock.parentElement.addEventListener.mock.calls.length).toBe(1);
  expect(mock.parentElement.removeEventListener.mock.calls.length).toBe(1);
});

it('renders a promise returning menu items properly', async () => {
  const mock = DEFAULT_MOCK();
  const promise = Promise.resolve(TEST_MENU_1);

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  const mockEvent = { preventDefault: jest.fn(), contextActions: [promise] };
  const handleContextMenu =
    mock.parentElement.addEventListener.mock.calls[0][1];
  handleContextMenu(mockEvent);
  await TestUtils.flushPromises();

  expect(tree.root.findAllByType('button').length).toBe(TEST_MENU_1.length);
  expect(tree).toMatchSnapshot();
});

it('renders an empty menu for a rejected promise', () => {
  const mock = DEFAULT_MOCK();
  const promise = Promise.reject();

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  const mockEvent = { preventDefault: jest.fn(), contextActions: [promise] };
  const handleContextMenu =
    mock.parentElement.addEventListener.mock.calls[0][1];
  handleContextMenu(mockEvent);

  expect(tree.root.findAllByType('button').length).toBe(0);
  expect(tree).toMatchSnapshot();
});

it('renders a menu from a promise returned from a function', () => {
  const mock = DEFAULT_MOCK();
  const promise = Promise.resolve([]);
  const fn = () => promise;

  const tree = TestRenderer.create(
    <div>
      <ContextMenuRoot />
    </div>,
    {
      createNodeMock: contextMenuMock.bind(this, mock),
    }
  );

  const mockEvent = { preventDefault: jest.fn(), contextActions: [fn] };
  const handleContextMenu =
    mock.parentElement.addEventListener.mock.calls[0][1];
  handleContextMenu(mockEvent);

  expect(tree.root.findAllByType('button').length).toBe(0);
  expect(tree).toMatchSnapshot();
});

it('sorts by title properly', () => {
  const actions = [{ title: 'b' }, { title: 'c' }, { title: 'a' }];
  const expectedResult = [actions[2], actions[0], actions[1]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});

it('sorts by order properly', () => {
  const actions = [
    { title: 'c', order: 40 },
    { title: 'b', order: 90 },
    { title: 'a', order: 10 },
  ];
  const expectedResult = [actions[2], actions[0], actions[1]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});

it('sorts groups properly', () => {
  const actions = [
    { title: 'c', order: 40, group: 10 },
    { title: 'b', order: 90, group: 10 },
    { title: 'a', order: 10, group: 30 },
  ];
  const expectedResult = [actions[0], actions[1], actions[2]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});

it('sorts mixed groups properly', () => {
  const actions = [
    { title: 'c', order: 40, group: 10 },
    { title: 'a', order: 10 },
    { title: 'b', order: 90, group: 10 },
  ];
  const expectedResult = [actions[1], actions[0], actions[2]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});

it('sorts groups/orders properly', () => {
  const actions = [
    { title: 'c', order: 104, group: 1 },
    { order: 10, group: 1 },
    { title: 'b', order: 105, group: 1 },
  ];
  const expectedResult = [actions[1], actions[0], actions[2]];
  const sortedActions = ContextActionUtils.sortActions(actions);
  expect(sortedActions).toEqual(expectedResult);
});
