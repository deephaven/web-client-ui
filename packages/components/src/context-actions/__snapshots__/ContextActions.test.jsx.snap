// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles the context menu event on the parent, renders an array of menu items 1`] = `
<div>
  <div
    className="context-menu-root"
  />
</div>
`;

exports[`renders a menu from a promise returned from a function 1`] = `
<div>
  <div
    className="context-menu-root active"
  >
    <div
      className="context-menu-container"
      data-dh-context-menu={true}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "left": NaN,
          "top": NaN,
        }
      }
      tabIndex="0"
    >
      <div
        className="loading"
      >
        <div
          className="loading-spinner fa-layers"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-circle-large-outline fa-w-16 text-white-50"
            data-icon="circle-large-outline"
            data-prefix="vs"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M306.816 70.88A185.856 185.856 0 0 0 256 64c-17.728 0 -34.624 2.336 -50.816 6.88l-0.192 0.064c-16.448 4.512 -31.68 10.944 -45.824 19.232A196.992 196.992 0 0 0 90.24 159.36l-0.064 0.128A190.944 190.944 0 0 0 70.72 205.12 191.552 191.552 0 0 0 64 256c0 17.76 2.24 34.72 6.72 50.912a193.6 193.6 0 0 0 49.536 84.832c11.84 11.68 24.768 21.664 38.912 30.08a195.2 195.2 0 0 0 45.92 19.488A192.64 192.64 0 0 0 256 448c17.76 0 34.72 -2.24 50.912 -6.72 16.48 -4.64 31.68 -11.136 45.632 -19.424l0.128 -0.064a197.12 197.12 0 0 0 69.152 -68.96 187.2 187.2 0 0 0 19.2 -45.824l0.096 -0.192A185.824 185.824 0 0 0 448 256c0 -17.728 -2.304 -34.624 -6.88 -50.816l-0.064 -0.192a184.704 184.704 0 0 0 -19.2 -45.536l-0.064 -0.128a188.8 188.8 0 0 0 -30.144 -38.72l-0.256 -0.256a188.864 188.864 0 0 0 -38.72 -30.144l-0.128 -0.064a184.704 184.704 0 0 0 -45.536 -19.2l-0.192 -0.064zm142.56 298.24a229.024 229.024 0 0 1 -80.512 80.256 222.912 222.912 0 0 1 -53.376 22.72A223.488 223.488 0 0 1 256 480a223.488 223.488 0 0 1 -59.52 -7.872 227.136 227.136 0 0 1 -53.568 -22.752 233.6 233.6 0 0 1 -45.28 -35.008 233.44 233.44 0 0 1 -35.008 -45.28 227.136 227.136 0 0 1 -22.72 -53.6A223.52 223.52 0 0 1 32 256c0 -20.576 2.624 -40.384 7.872 -59.52a222.976 222.976 0 0 1 22.752 -53.344 228.992 228.992 0 0 1 80.288 -80.512 220.64 220.64 0 0 1 53.6 -22.528A217.856 217.856 0 0 1 256 32a217.6 217.6 0 0 1 59.52 8.096 220.768 220.768 0 0 1 98.656 57.76 220.896 220.896 0 0 1 57.728 98.656C477.312 215.616 480 235.424 480 256s-2.688 40.384 -8.096 59.52a220.8 220.8 0 0 1 -22.528 53.568z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-loading fa-w-16 fa-spin text-primary"
            data-icon="loading"
            data-prefix="vs"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M445.344 224A192.064 192.064 0 0 0 66.656 224H34.272a224.064 224.064 0 0 1 443.456 0h-32.384z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders a promise returning menu items properly 1`] = `
<div>
  <div
    className="context-menu-root active"
  >
    <div
      className="context-menu-container"
      data-dh-context-menu={true}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "left": NaN,
          "top": NaN,
        }
      }
      tabIndex="0"
    >
      <div
        className="context-menu-item"
      >
        <button
          className="btn-context-menu"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseMove={[Function]}
          title=""
          type="button"
        >
          <div
            className="btn-context-menu-wrapper"
          >
            <span
              className="icon"
            />
            <span
              className="title"
            >
              Test1
            </span>
          </div>
        </button>
      </div>
      <div
        className="context-menu-item"
      >
        <button
          className="btn-context-menu"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseMove={[Function]}
          title=""
          type="button"
        >
          <div
            className="btn-context-menu-wrapper"
          >
            <span
              className="icon"
            />
            <span
              className="title"
            >
              Test2
            </span>
          </div>
        </button>
      </div>
      <div
        className="context-menu-item"
      >
        <button
          className="btn-context-menu"
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseMove={[Function]}
          title=""
          type="button"
        >
          <div
            className="btn-context-menu-wrapper"
          >
            <span
              className="icon"
            />
            <span
              className="title"
            >
              Test3
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders an empty menu for a rejected promise 1`] = `
<div>
  <div
    className="context-menu-root active"
  >
    <div
      className="context-menu-container"
      data-dh-context-menu={true}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="menuitem"
      style={
        Object {
          "left": NaN,
          "top": NaN,
        }
      }
      tabIndex="0"
    >
      <div
        className="loading"
      >
        <div
          className="loading-spinner fa-layers"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-circle-large-outline fa-w-16 text-white-50"
            data-icon="circle-large-outline"
            data-prefix="vs"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M306.816 70.88A185.856 185.856 0 0 0 256 64c-17.728 0 -34.624 2.336 -50.816 6.88l-0.192 0.064c-16.448 4.512 -31.68 10.944 -45.824 19.232A196.992 196.992 0 0 0 90.24 159.36l-0.064 0.128A190.944 190.944 0 0 0 70.72 205.12 191.552 191.552 0 0 0 64 256c0 17.76 2.24 34.72 6.72 50.912a193.6 193.6 0 0 0 49.536 84.832c11.84 11.68 24.768 21.664 38.912 30.08a195.2 195.2 0 0 0 45.92 19.488A192.64 192.64 0 0 0 256 448c17.76 0 34.72 -2.24 50.912 -6.72 16.48 -4.64 31.68 -11.136 45.632 -19.424l0.128 -0.064a197.12 197.12 0 0 0 69.152 -68.96 187.2 187.2 0 0 0 19.2 -45.824l0.096 -0.192A185.824 185.824 0 0 0 448 256c0 -17.728 -2.304 -34.624 -6.88 -50.816l-0.064 -0.192a184.704 184.704 0 0 0 -19.2 -45.536l-0.064 -0.128a188.8 188.8 0 0 0 -30.144 -38.72l-0.256 -0.256a188.864 188.864 0 0 0 -38.72 -30.144l-0.128 -0.064a184.704 184.704 0 0 0 -45.536 -19.2l-0.192 -0.064zm142.56 298.24a229.024 229.024 0 0 1 -80.512 80.256 222.912 222.912 0 0 1 -53.376 22.72A223.488 223.488 0 0 1 256 480a223.488 223.488 0 0 1 -59.52 -7.872 227.136 227.136 0 0 1 -53.568 -22.752 233.6 233.6 0 0 1 -45.28 -35.008 233.44 233.44 0 0 1 -35.008 -45.28 227.136 227.136 0 0 1 -22.72 -53.6A223.52 223.52 0 0 1 32 256c0 -20.576 2.624 -40.384 7.872 -59.52a222.976 222.976 0 0 1 22.752 -53.344 228.992 228.992 0 0 1 80.288 -80.512 220.64 220.64 0 0 1 53.6 -22.528A217.856 217.856 0 0 1 256 32a217.6 217.6 0 0 1 59.52 8.096 220.768 220.768 0 0 1 98.656 57.76 220.896 220.896 0 0 1 57.728 98.656C477.312 215.616 480 235.424 480 256s-2.688 40.384 -8.096 59.52a220.8 220.8 0 0 1 -22.528 53.568z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-loading fa-w-16 fa-spin text-primary"
            data-icon="loading"
            data-prefix="vs"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M445.344 224A192.064 192.064 0 0 0 66.656 224H34.272a224.064 224.064 0 0 1 443.456 0h-32.384z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders without crashing 1`] = `
<div>
  <div
    className="context-menu-root"
  />
</div>
`;
